File: core-vm/java/java/io/RandomAccessFile.java
/**************************************************************************
* Copyright (c) 2001 by Punch Telematix. All rights reserved.             *
*                                                                         *
* Redistribution and use in source and binary forms, with or without      *
* modification, are permitted provided that the following conditions      *
* are met:                                                                *
* 1. Redistributions of source code must retain the above copyright       *
*    notice, this list of conditions and the following disclaimer.        *
* 2. Redistributions in binary form must reproduce the above copyright    *
*    notice, this list of conditions and the following disclaimer in the  *
*    documentation and/or other materials provided with the distribution. *
* 3. Neither the name of Punch Telematix nor the names of                 *
*    other contributors may be used to endorse or promote products        *
*    derived from this software without specific prior written permission.*
*                                                                         *
* THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED          *
* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.    *
* IN NO EVENT SHALL PUNCH TELEMATIX OR OTHER CONTRIBUTORS BE LIABLE       *
* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR            *
* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF    *
* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR         *
* BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,   *
* WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE    *
* OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN  *
* IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.                           *
**************************************************************************/



package java.io;

public class RandomAccessFile implements DataOutput, DataInput {

  private FileDescriptor fd;
  private native int createFromString(String name, int nativeMode);

  private byte[] buffer = new byte[64];

  public RandomAccessFile(String name, String mode) throws FileNotFoundException {
    this(new File(name), mode);
  }

  public RandomAccessFile(File file, String mode) throws FileNotFoundException {
    int nativeMode;
    if(mode.equals("r")) {
      nativeMode = 0; 
    } else if (mode.equals("rw")) {
      nativeMode = 1; 
    } else if (mode.equals("rwd")){
      nativeMode = 2;
    } else if (mode.equals("rws")) {
      nativeMode = 3;
    } else {
      throw new IllegalArgumentException(mode);
    }    
    if(createFromString(file.getAbsolutePath(), nativeMode) != 0) {
      throw new FileNotFoundException("" + file + " (" + mode + ")");
    }
  }

  public final FileDescriptor getFD()
    throws IOException {
    return fd;
  }

  public native int read() 
    throws IOException;

  private native int readIntoBuffer(byte[]b, int off, int len);

  public int read(byte[] b, int off, int len) 
    throws IOException, NullPointerException, ArrayIndexOutOfBoundsException {
    
    if(off<0 || len<0 || off > b.length - len)
      throw new ArrayIndexOutOfBoundsException();
    if(len == 0) {
      return 0;
    }
    return readIntoBuffer(b, off, len);
  }
 
  public int read(byte[] b)
    throws IOException, NullPointerException {
    if(b.length == 0) return 0; 
    return readIntoBuffer(b, 0, b.length);
  }

  public final void readFully(byte[] b)
    throws IOException, NullPointerException, ArrayIndexOutOfBoundsException, EOFException {
    int rc;
    int l = 0;

    while (l < b.length) {
      rc = read(b, l, b.length - l);
      if (rc == -1) {
        throw new EOFException();
      }
      l += rc;
    }
  }

  public final void readFully(byte[] b, int off, int len)
    throws IOException, NullPointerException, EOFException {
    int rc;
    int l = 0;

    while (l < len) {
      wonka.vm.Etc.woempa(7, "calling read(" + b + ", " + (off + l) + ", " + (len - l));
      rc = read(b, off + l, len - l);
      if (rc == -1) {
        throw new EOFException();
      }
      l += rc;
    }
  }

  public native int skipBytes(int n)
    throws IOException;

  public native void write(int b)
    throws IOException;

  private native void writeFromBuffer(byte[] b, int off, int len)
    throws IOException;

  public void write(byte[] b)
    throws IOException, NullPointerException {

    if(b==null) throw new NullPointerException();
    writeFromBuffer(b, 0, b.length);
  }

  public void write(byte[] b, int off, int len)
    throws IOException, NullPointerException, ArrayIndexOutOfBoundsException {

    if(off<0 || len<0 || off > b.length - len)
      throw new ArrayIndexOutOfBoundsException();

    writeFromBuffer(b, off,len);
  }

  public native long getFilePointer()
    throws IOException;

  public native void seek(long pos)
    throws IOException;

  public native long length()
    throws IOException;

  public native void setLength(long NewLength)
    throws IOException;

  public native void close()
    throws IOException;

  public final boolean readBoolean() throws IOException {
    this.readFully(buffer, 0, 1);

    if(buffer[0] == 0) 
      return false;
    else
      return true;
  }

  public final byte readByte() throws IOException {
    this.readFully(buffer, 0, 1);
    return buffer[0];
  }

  public final int readUnsignedByte() throws IOException {
    this.readFully(buffer, 0, 1);
    return (0 | buffer[0]);
  }

  public final short readShort() throws IOException {
    this.readFully(buffer, 0, 2);
    return (short)((buffer[0] << 8) | buffer[1]);
  }

  public final int readUnsignedShort() throws IOException {
    this.readFully(buffer, 0, 2);
    return ((buffer[0] << 8) | buffer[1]);
  }

  public final char readChar() throws IOException {
    this.readFully(buffer, 0, 2);
    return (char)((buffer[0] << 8) | buffer[1]);
  }

  public final int readInt() throws IOException {
    this.readFully(buffer, 0, 4);
    return (((buffer[0] & 0xFF) << 24) | ((buffer[1] & 0xFF) << 16) | ((buffer[2] & 0xFF) << 8) | (buffer[3] & 0xFF));
  }

  public final long readLong() throws IOException {
    this.readFully(buffer, 0, 8);
    return (((long)buffer[0] & 0x0ff) << 56) | (((long)buffer[1] & 0x0ff)<< 48) | 
           (((long)buffer[2] & 0x0ff) << 40) | (((long)buffer[3] & 0x0ff) << 32) |
           (((long)buffer[4] & 0x0ff) << 24) | (((long)buffer[5] & 0x0ff) << 16) |
           (((long)buffer[6] & 0x0ff) <<  8) | (buffer[7] & 0x0ff);
  }

  public final float readFloat() throws IOException {
    return Float.intBitsToFloat(this.readInt());
  }

  public final double readDouble() throws IOException {
    return Double.longBitsToDouble(this.readLong());
  }

  public final void writeBoolean(boolean v) throws IOException {
    if(v) this.write(1); else this.write(0);
  }

  public final void writeByte(int v) throws IOException {
    this.write(v);
  }
  
  public final void writeShort(int v) throws IOException {
    this.write(v >> 8);
    this.write(v & 0x00FF);
  }

  public final void writeChar(int v) throws IOException {
    this.write((v >> 8) & 0x000000FF);
    this.write(v & 0x000000FF);
  }

  public final void writeInt(int v)throws IOException {
    this.write((v >> 24) & 0x000000FF);
    this.write((v >> 16) & 0x000000FF);
    this.write((v >> 8) & 0x000000FF);
    this.write(v & 0x000000FF);
  }

  public final void writeLong(long v) throws IOException {
    this.write((int)((v >> 56) & 0x000000FF));
    this.write((int)((v >> 48) & 0x000000FF));
    this.write((int)((v >> 40) & 0x000000FF));
    this.write((int)((v >> 32) & 0x000000FF));
    this.write((int)((v >> 24) & 0x000000FF));
    this.write((int)((v >> 16) & 0x000000FF));
    this.write((int)((v >> 8) & 0x000000FF));
    this.write((int)(v & 0x000000FF));
  }

  public final void writeFloat(float v) throws IOException {
    this.writeInt(Float.floatToIntBits(v));
  }

  public final void writeDouble(double v) throws IOException {
    this.writeLong(Double.doubleToLongBits(v));
  }

  public final String readLine() throws IOException {
    StringBuffer sb = new StringBuffer();
    long pos = getFilePointer();
    long l = length();
    char ch;

    while (pos < l) {
      ch = (char)(read() & 255);
      if (ch == 13) { // CR
        if (pos < l - 1) {
          if (read() != 10) { // LF
            seek(pos + 1);
          }
        }
        break;
      }
      else if (ch == 10) { // LF
        break;
      }
      sb.append(ch);
      ++pos;
    }

    return sb.toString();
  }

  public final String readUTF() 
    throws IOException {
    int UTFlength = this.readUnsignedShort();
    if (UTFlength < 0){
      throw new UTFDataFormatException();
    }
    byte[] bytes = new byte[UTFlength];
    this.readFully(bytes);
    return new String(bytes,"UTF8");
  }

  public final void writeBytes(String s) throws IOException {
    char c[] = s.toCharArray();
    for(int i=0; i < c.length; i++) {
      write((byte)c[i]);
    }
  }

  public final void writeChars(String s) throws IOException {
    char c[] = s.toCharArray();
    for(int i=0; i < c.length; i++) {
      writeChar(c[i]);
    }
  }

  public final void writeUTF(String str) throws IOException {
    byte []b = str.getBytes("UTF8");
    this.writeShort(b.length);
    this.write(b, 0, b.length);
  }

  protected void finalize() throws Throwable {
    if(fd != null) {
      close();
    }
  }
}

File: Mika/branches/java.math/core-vm/src/native/java/io/RandomAccessFile.c
/**************************************************************************
* Parts copyright (c) 2001, 2002, 2003 by Punch Telematix.                *
* All rights reserved.                                                    *
* Parts copyright (c) 2007, 2008 by Chris Gray, /k/ Embedded Java         *
* Solutions. All rights reserved.                                         *
*                                                                         *
* Redistribution and use in source and binary forms, with or without      *
* modification, are permitted provided that the following conditions      *
* are met:                                                                *
* 1. Redistributions of source code must retain the above copyright       *
*    notice, this list of conditions and the following disclaimer.        *
* 2. Redistributions in binary form must reproduce the above copyright    *
*    notice, this list of conditions and the following disclaimer in the  *
*    documentation and/or other materials provided with the distribution. *
* 3. Neither the name of Punch Telematix or of /k/ Embedded Java Solutions*
*    nor the names of other contributors may be used to endorse or promote*
*    products derived from this software without specific prior written   *
*    permission.                                                          *
*                                                                         *
* THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED          *
* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.    *
* IN NO EVENT SHALL PUNCH TELEMATIX, /K/ EMBEDDED JAVA SOLUTIONS OR OTHER *
* CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,   *
* EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,     *
* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR      *
* PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF  *
* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING    *
* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS      *
* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.            *
**************************************************************************/

#include "core-classes.h"
#include "exception.h"
#include "fields.h"
#include "loading.h"
#include "vfs.h"
#include "wstrings.h"

/*
 * Class:     RandomAccessFile
 * Method:    createFromString
 * Signature: (Ljava/lang/String;Ljava/lang/String;)V
 */
w_int Java_RandomAccessFile_createFromString (JNIEnv *env, w_instance thisRAF, w_instance path, int mode) {
  w_thread thread = JNIEnv2w_thread(env);
  w_string pathname_string;
  char *pathname;
  const char  *fmode = "r+";
  int         openmode;
  int         fdesc;
  vfs_FILE    *file;
  w_instance  fd_obj;
  struct vfs_STAT statbuf;
  int rc;

  if (mustBeInitialized(clazzFileDescriptor) == CLASS_LOADING_FAILED) {
    return 0;
  }

  pathname_string = String2string(path);
  pathname = (char*)string2UTF8(pathname_string, NULL) + 2;	  
  switch(mode) {
    case 0:
      openmode = VFS_O_RDONLY;
      fmode = "r";
      break;
    case 1:
      openmode = VFS_O_RDWR | VFS_O_CREAT;
      break;
    case 2:
      openmode = VFS_O_RDWR | VFS_O_CREAT | VFS_O_DIRECT;
      break;
    case 3:
      openmode = VFS_O_RDWR | VFS_O_CREAT | VFS_O_SYNC;
      break;
    default:
     openmode = 0;
  }

  rc = vfs_stat(pathname, &statbuf);

  switch(mode) {
    case 0:
      if (rc != 0 || !VFS_S_ISREG(statbuf.st_mode) || (((statbuf.st_mode & VFS_S_IRWXU) & VFS_S_IRUSR) != VFS_S_IRUSR)) {
        return 1;
      }
      break;

    default:
      if (rc == 0 && (!VFS_S_ISREG(statbuf.st_mode) || (((statbuf.st_mode & VFS_S_IRWXU) & VFS_S_IWUSR) != VFS_S_IWUSR))) {
        return 1;
      }
  }

  fdesc = vfs_open(pathname, openmode, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);

  releaseMem(pathname - 2);

  if(fdesc == -1) {
    return 1;
  } 

  file = vfs_fdopen(fdesc, fmode);

  if (file == NULL) {
    return 1;
  } else {
    enterUnsafeRegion(thread);
    fd_obj = allocInstance(thread, clazzFileDescriptor);
    enterSafeRegion(thread);
    if(fd_obj != NULL) {
      setReferenceField(thisRAF, fd_obj, F_RandomAccessFile_fd);
      setBooleanField(fd_obj, F_FileDescriptor_validFD, 1);
      setWotsitField(fd_obj, F_FileDescriptor_fd, file);
      setReferenceField(fd_obj, path, F_FileDescriptor_fileName);
    }
  }
  return 0;
}

/*
 * Class:     RandomAccessFile
 * Method:    read
 * Signature: ()I
 */
w_int Java_RandomAccessFile_read (JNIEnv *env, w_instance thisRAF) {

  w_instance  fd_obj;
  vfs_FILE    *file;
  w_int       result = -1;

  fd_obj = getReferenceField(thisRAF, F_RandomAccessFile_fd);
  file = getWotsitField(fd_obj, F_FileDescriptor_fd);
  
  if(file == NULL) {
    throwNullPointerException(JNIEnv2w_thread(env));
    result = 0;
  } else {
    result = vfs_fgetc(file);
    //if(!vfs_feof(file)) result = vfs_fgetc(file);
  }

  return result;
}

/*
 * Class:     RandomAccessFile
 * Method:    readIntoBuffer
 * Signature: ([BII)I
 */
w_int Java_RandomAccessFile_readIntoBuffer (JNIEnv *env, w_instance thisRAF, w_instance buffer, w_int offset, w_int length) {

  w_instance  fd_obj;
  vfs_FILE    *file;
  w_int       result;
  w_sbyte     *bytes;
  w_sbyte     *data;

  if ((offset < 0) || (offset > instance2Array_length(buffer) - length)) {
    throwArrayIndexOutOfBoundsException(JNIEnv2w_thread(env));

    return -1;
  }

  bytes = instance2Array_byte(buffer);
  fd_obj = getReferenceField(thisRAF, F_RandomAccessFile_fd);
  file = getWotsitField(fd_obj, F_FileDescriptor_fd);
  
  if(file == NULL) {
    throwNullPointerException(JNIEnv2w_thread(env));
    result = 0;
  } else {
    data = bytes + offset;
    result = vfs_fread(data, 1, (w_word)length, file);
    if(result == 0) result = -1;
  }

  return result;
}

/*
 * Class:     RandomAccessFile
 * Method:    skipBytes
 * Signature: (I)I
 */
w_int Java_RandomAccessFile_skipBytes (JNIEnv *env, w_instance thisRAF, w_int n) {

  w_instance  fd_obj;
  vfs_FILE    *file;
  w_long       result = 0;
  w_long       prev_pos;

  fd_obj = getReferenceField(thisRAF, F_RandomAccessFile_fd);
  file = getWotsitField(fd_obj, F_FileDescriptor_fd);
  
  if(file == NULL) {
    throwNullPointerException(JNIEnv2w_thread(env));
  } else {
    w_string pathname_string;
    const char *pathname;
    w_instance path;
    w_int pathlen;
    struct vfs_STAT statbuf;
    w_long size = 0;

    path = getReferenceField(fd_obj, F_FileDescriptor_fileName);
    pathname_string = String2string(path);
    pathname = (char*)string2UTF8(pathname_string, &pathlen) + 2;

    if(vfs_stat(pathname, &statbuf) != -1) {
      size = statbuf.st_size;
    }

    releaseMem(pathname - 2);


    prev_pos = vfs_ftell(file);

    if(n > (size - prev_pos)) {
      n = size - prev_pos;
    }
     
    result = vfs_fseek(file, (long)n, SEEK_CUR);

    if(result == 0) {
      result = vfs_ftell(file) - prev_pos;
    } else {
      throwIOException(JNIEnv2w_thread(env));
    }
  }

  return result;
}

/*
 * Class:     RandomAccessFile
 * Method:    write
 * Signature: (I)V
 */
void Java_RandomAccessFile_write (JNIEnv *env, w_instance thisRAF, w_int oneByte) {

  w_instance  fd_obj;
  vfs_FILE    *file;

  fd_obj = getReferenceField(thisRAF, F_RandomAccessFile_fd);
  file = getWotsitField(fd_obj, F_FileDescriptor_fd);
  
  if(file == NULL) {
    throwNullPointerException(JNIEnv2w_thread(env));
  } else {
    vfs_fputc(oneByte, file);
  }
}

/*
 * Class:     RandomAccessFile
 * Method:    writeFromBuffer
 * Signature: ([BII)V
 */
void Java_RandomAccessFile_writeFromBuffer (JNIEnv *env, w_instance thisRAF, w_instance buffer, w_int offset, w_int length) {

  w_instance  fd_obj;
  vfs_FILE    *file;
  w_sbyte     *bytes;
  w_sbyte     *data;

  if ((offset < 0) || (offset > instance2Array_length(buffer) - length)) {
    throwArrayIndexOutOfBoundsException(JNIEnv2w_thread(env));

    return;
  }

  bytes = instance2Array_byte(buffer);
  fd_obj = getReferenceField(thisRAF, F_RandomAccessFile_fd);
  file = getWotsitField(fd_obj, F_FileDescriptor_fd);
  
  if(file == NULL) {
    throwNullPointerException(JNIEnv2w_thread(env));
  } else {
    data = bytes + offset;
    vfs_fwrite(data, 1, (w_word)length, file);
    vfs_fflush(file);
  }
}

/*
 * Class:     RandomAccessFile
 * Method:    getFilePointer
 * Signature: ()J
 */
w_long Java_RandomAccessFile_getFilePointer (JNIEnv *env, w_instance thisRAF) {

  w_instance  fd_obj;
  vfs_FILE    *file;
  w_long       result = -1;

  fd_obj = getReferenceField(thisRAF, F_RandomAccessFile_fd);
  file = getWotsitField(fd_obj, F_FileDescriptor_fd);
  
  if(file == NULL) {
    throwNullPointerException(JNIEnv2w_thread(env));
    result = 0;
  } else {
    result = vfs_ftell(file);
  }

  return result;
}

/*
 * Class:     RandomAccessFile
 * Method:    seek
 * Signature: (J)V
 */
void Java_RandomAccessFile_seek (JNIEnv *env, w_instance thisRAF, w_long pos) {

  w_instance  fd_obj;
  vfs_FILE    *file;

  fd_obj = getReferenceField(thisRAF, F_RandomAccessFile_fd);
  file = getWotsitField(fd_obj, F_FileDescriptor_fd);
  
  if(file == NULL) {
    throwNullPointerException(JNIEnv2w_thread(env));
  } else {
    if(vfs_fseek(file, (long)pos, SEEK_SET) == -1) {
      throwIOException(JNIEnv2w_thread(env));
    }
  }
}

/*
 * Class:     RandomAccessFile
 * Method:    length
 * Signature: ()J
 */
w_long Java_RandomAccessFile_length (JNIEnv *env, w_instance thisRAF) {

  w_instance fd_obj;
  w_string pathname_string;
  const char *pathname;
  w_instance path;
  w_int pathlen;
  struct vfs_STAT statbuf;
  w_long result = 0;
  
  fd_obj = getReferenceField(thisRAF, F_RandomAccessFile_fd);
  path = getReferenceField(fd_obj, F_FileDescriptor_fileName);
  pathname_string = String2string(path);
  pathname = (char*)string2UTF8(pathname_string, &pathlen) + 2;	  

  if (vfs_stat(pathname, &statbuf) == -1) {
    throwIOException(JNIEnv2w_thread(env));
  }
  result = statbuf.st_size;
  woempa(7, "vfs_stat(%s, %p) returned length %d\n", pathname, &statbuf, result);
  releaseMem(pathname - 2);

  return result;
}

/*
 * Class:     RandomAccessFile
 * Method:    setLength
 * Signature: (J)V
 */
void Java_RandomAccessFile_setLength (JNIEnv *env, w_instance thisRAF, w_long newlen) {

  w_instance  fd_obj;
  vfs_FILE    *file;

  fd_obj = getReferenceField(thisRAF, F_RandomAccessFile_fd);
  file = getWotsitField(fd_obj, F_FileDescriptor_fd);
  
  if(file == NULL) {
    throwNullPointerException(JNIEnv2w_thread(env));
  } else {
    w_string pathname_string;
    const char *pathname;
    w_instance path;
    w_int pathlen;
    w_long oldptr;

    path = getReferenceField(fd_obj, F_FileDescriptor_fileName);
    pathname_string = String2string(path);
    pathname = (char*)string2UTF8(pathname_string, &pathlen) + 2;

    oldptr = vfs_ftell(file);

    if(vfs_truncate(pathname, newlen) == -1) {
      throwIOException(JNIEnv2w_thread(env));
    }

    releaseMem(pathname - 2);

    if(newlen < oldptr && vfs_fseek(file, newlen, SEEK_SET) == -1) {
      throwIOException(JNIEnv2w_thread(env));
    }
  }
}

/*
 * Class:     RandomAccessFile
 * Method:    close
 * Signature: ()V
 */
void Java_RandomAccessFile_close (JNIEnv *env, w_instance thisRAF) { 
  w_instance  fd_obj;
  vfs_FILE    *file;

  fd_obj = getReferenceField(thisRAF, F_RandomAccessFile_fd);

  if(fd_obj == NULL) {
    throwNullPointerException(JNIEnv2w_thread(env));
  } else {
    
    file = getWotsitField(fd_obj, F_FileDescriptor_fd);
  
    if(file) {  
      vfs_fclose(file);
      clearWotsitField(fd_obj, F_FileDescriptor_fd);
    }
    
  }
  
}
